### PR TITLE
feat(types): add mana allotment struct

### LINKED ISSUE #913: Add mana `Allotment` struct - rust
https://github.com/iotaledger/tips-draft/blob/tip45/tips/TIP-0045/tip-0045.md#serialized-layout

https://github.com/iotaledger/tips-draft/blob/tip45/tips/TIP-0045/tip-0045.md#allotments

### PR DESCRIPTION
**Pull request by Alex6323**
# Description of change

This adds a struct to represent mana allotment in a `RegularTransactionEssence`.

## Links to any relevant issues

#913 

## Type of change

Choose a type of change, and delete any options that are not relevant.

- Enhancement (a non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: sdk/src/types/block/mana/allotment.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    use serde::{Deserialize, Serialize};
+
+    use super::*;
+    use crate::types::{block::Error, TryFromDto, ValidationParams};
+
+    /// TODO
+    #[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
+    pub struct AllotmentDto {
+        pub account_id: AccountId,
+        #[serde(with = "crate::utils::serde::string")]
```
Comment (kwek20) [inline sdk/src/types/block/mana/allotment.rs]: can you import this, were doing the with = "string" everywhere

File: sdk/src/types/block/mana/allotment.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+};
+
+use crate::types::block::{output::AccountId, Error};
+
+/// The maximum number of allotments of a transaction.
+pub const ALLOTMENT_COUNT_MAX: u16 = 128;
+/// The range of valid numbers of allotments of a transaction.
+pub const ALLOTMENT_COUNT_RANGE: RangeInclusive<u16> = 1..=ALLOTMENT_COUNT_MAX; // [1..128]
+
+/// TODO
```
Comment (kwek20) [inline sdk/src/types/block/mana/allotment.rs]: Whats the todo?

File: sdk/src/types/block/mana/allotment.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    error::{UnpackError, UnpackErrorExt},
+    packer::Packer,
+    unpacker::Unpacker,
+    Packable,
+};
+
+use crate::types::block::{output::AccountId, Error};
+
+/// The maximum number of allotments of a transaction.
+pub const ALLOTMENT_COUNT_MAX: u16 = 128;
```
Comment (DaughterOfMars) [inline sdk/src/types/block/mana/allotment.rs]: Can we put these on an `Allotments` type like `Features` or `UnlockConditions`?

File: sdk/src/types/block/mana/allotment.rs (hunk 1 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/mana/allotment.rs]: We don't need a DTO here at all

Comment (DaughterOfMars) [reply sdk/src/types/block/mana/allotment.rs]: Actually, if we need mana supply in `TryFromDto`, I guess we will.

File: sdk/src/types/block/mana/allotment.rs (hunk 2 above)
Comment (Alex6323) [reply sdk/src/types/block/mana/allotment.rs]: hmm, the docs? 🙂

File: sdk/src/types/block/mana/mod.rs [hunk 4]
```diff
@@ -3,4 +3,92 @@
...
+        verify_allotments::<true>(&allotments, &())?;
+
+        Ok(Self(allotments))
+    }
+
+    /// Gets a reference to an [`Allotment`], if one exists, using an [`AccountId`].
+    #[inline(always)]
+    pub fn get(&self, account_id: &AccountId) -> Option<&Allotment> {
+        self.0
+            .binary_search_by(|a| a.account_id().cmp(account_id))
```
Comment (Thoralf-M) [inline sdk/src/types/block/mana/mod.rs]: `binary_search_by()` can't be used here since we don't make sure that it's ordered when we construct it?

Comment (DaughterOfMars) [reply sdk/src/types/block/mana/mod.rs]: Ah, good point

File: sdk/src/types/block/mana/mod.rs [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    /// Gets a reference to an [`Allotment`], if one exists, using an [`AccountId`].
+    #[inline(always)]
+    pub fn get(&self, account_id: &AccountId) -> Option<&Allotment> {
+        self.0
+            .iter()
+            .position(|a| a.account_id() == account_id)
+            // PANIC: indexation is fine since the index has been found.
+            .map(|index| &self.0[index])
```
Comment (Thoralf-M) [inline sdk/src/types/block/mana/mod.rs:93]: Why not just `.find(|a| a.account_id() == account_id)`?

File: sdk/src/types/block/mana/allotment.rs [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    impl<'de> Deserialize<'de> for Allotment {
+        fn deserialize<D>(d: D) -> Result<Self, D::Error>
+        where
+            D: serde::Deserializer<'de>,
+        {
+            #[derive(Deserialize)]
+            #[serde(rename_all = "camelCase")]
+            struct AllotmentDto {
+                account_id: AccountId,
+                #[cfg_attr(feature = "serde", serde(with = "string"))]
```
Comment (kwek20) [inline sdk/src/types/block/mana/allotment.rs]: already checked for feature

File: sdk/src/types/block/mana/mod.rs (hunk 5 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/mana/mod.rs:93]: Doesn't exist on boxed slices

File: sdk/src/types/block/mana/allotment.rs (hunk 6 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/mana/allotment.rs]: oop

**Review by Thoralf-M — APPROVED** [review]

**Review by kwek20 — APPROVED** [review]

File: sdk/src/types/block/mana/mod.rs [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    type IntoIter = alloc::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        Vec::from(Into::<Box<[Allotment]>>::into(self.0)).into_iter()
+    }
+}
+
+impl Allotments {
+    /// The minimum number of allotments of a transaction.
+    pub const COUNT_MIN: u16 = 1;
```
Comment (kwek20) [inline sdk/src/types/block/mana/mod.rs:70]: actually this is 0

Comment (kwek20) [reply sdk/src/types/block/mana/mod.rs:70]: `Allottments Count must be 0 < x ≤ Max Allottments Count.`

Comment (DaughterOfMars) [reply sdk/src/types/block/mana/mod.rs:70]: No, it's greater than 0 :)

Comment (kwek20) [reply sdk/src/types/block/mana/mod.rs:70]: uhg stupid fancy notations :P1º

File: sdk/src/types/block/mana/mod.rs [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+use hashbrown::HashSet;
+use packable::{bounded::BoundedU16, prefix::BoxedSlicePrefix, Packable};
+
+pub use self::allotment::Allotment;
+use super::{output::AccountId, Error};
+
+/// The number of bits that a given mana value can use, excluding the sign bit.
+pub const MANA_BITS: u64 = 63;
+/// Equivalent to `2^MANA_BITS - 1`
+pub const MAX_THEORETICAL_MANA: u64 = u64::MAX >> 1;
```
Comment (abdulmth) [inline sdk/src/types/block/mana/mod.rs:19]: I wonder about the name, why not `MAX_MANA_SUPPLY`?

**Review by abdulmth — APPROVED** [review]

File: sdk/src/types/block/mana/mod.rs (hunk 8 above)
Comment (DaughterOfMars) [reply sdk/src/types/block/mana/mod.rs:19]: Well, that's kind of wrong since there is no supply.

**Review by qrayven — APPROVED** [review]

**Pull request merged**